File: Python_Books/算法图解/grokking_algorithms_COPY/01_introduction_to_algorithms/javascript/01_binary_search.js
'use strict';

function binary_search(list, item) {
  let low = 0;
  let high = list.length - 1;
  
  while (low <= high) {
    let mid = Math.floor((low + high) / 2);
    let guess = list[mid];
    if (guess === item) {
      return mid;
    }
    if (guess > item) {
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }
  
  return null;
}

const my_list = [1, 3, 5, 7, 9];

console.log(binary_search(my_list, 3)); // 1
console.log(binary_search(my_list, -1)); // null

File: Python_Books/算法图解/grokking_algorithms_COPY/01_introduction_to_algorithms/lua/01_binary_search.lua
function binary_search(array, value)
    -- Lua arrays start with 1
    local low, high = 1, #array
    
    while low <= high do
        local mid = math.floor((low + high) / 2)
        local guess = array[mid]
        
        if guess == value then
            return mid
        elseif guess > value then
            high = mid - 1
        else
            low = mid + 1
        end
    end
    
    return nil
end

local my_array = {1, 3, 5, 7, 9}

print(binary_search(my_array, 3)) -- => 2
print(binary_search(my_array, -1)) -- => nil